migrations/20231128_jellyfish_migration/utils/data_migration.go

```diff
@@ -0,0 +1,386 @@
+	}
+	if checker == nil {
+		err = errors.Join(err, errors.New("missing required mongo checker"))
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	m := &DataMigration{
+		ctx:                  ctx,
+		dataC:                dataC,
+		mongoInstanceChecker: checker,
+		settings:             settings,
+		updates:              []mongo.WriteModel{},
+		groupedErrors:        groupedErrors{},
+		groupedDiffs:         map[string][]UpdateData{},
+		errorsCount:          0,
+		updatedCount:         0,
+		fetchedCount:         0,
+		startedAt:            time.Now(),
+	}
+	if lastID != nil {
+		m.lastUpdatedID = *lastID
+	}
+	return m, nil
+}
+
+func (m *DataMigration) Initialize() error {
+	if err := m.mongoInstanceChecker.CheckFreeSpace(m.ctx, m.dataC); err != nil {
+		return err
+	}
+	if err := m.mongoInstanceChecker.SetWriteBatchSize(m.ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *DataMigration) GetCtx() context.Context {
+	return m.ctx
+}
+
+func (m *DataMigration) GetSettings() Settings {
+	m.settings.WriteBatchSize = m.mongoInstanceChecker.GetWriteBatchSize()
+	return *m.settings
+}
+
+func (m *DataMigration) GetDataCollection() *mongo.Collection {
+	return m.dataC
+}
+
+func (m *DataMigration) Execute(
+	queryFn DataMigrationQueryFn,
+	updateFn DataMigrationUpdateFn) error {
+	for queryFn(m) {
+		count, err := updateFn(m)
+		if err != nil {
+			m.writeErrors(nil)
+			return err
+		}
+		m.updatesApplied(count)
+		if m.completeUpdates() {
+			break
+		}
+		m.writeErrors(nil)
+		m.writeAudit(nil)
+	}
+	m.GetStats().report()
+	m.writeErrors(nil)
+	m.writeAudit(nil)
+	return nil
+}
+
+func (d UpdateData) getMongoUpdates(rollback bool, rollbackSectionName string) []mongo.WriteModel {
+	updates := []mongo.WriteModel{}
+	for _, u := range d.Apply {
+		updateOp := mongo.NewUpdateOneModel()
+		updateOp.Filter = d.Filter
+		updateOp.SetUpdate(u)
+		updates = append(updates, updateOp)
+	}
+	updateOp := mongo.NewUpdateOneModel()
+	updateOp.Filter = d.Filter
+	if !rollback && len(d.Revert) > 0 {
+		updateOp.SetUpdate(bson.M{"$set": bson.M{rollbackSectionName: d.Revert}})
+	} else if rollback {
+		updateOp.SetUpdate(bson.M{"$unset": bson.M{rollbackSectionName: ""}})
+	}
+	updates = append(updates, updateOp)
+	return updates
+}
+
+func (m *DataMigration) SetUpdates(data UpdateData) {
+	m.groupedDiffs[data.ItemType] = append(m.groupedDiffs[data.ItemType], data)
+	m.updates = append(m.updates, data.getMongoUpdates(m.settings.Rollback, m.settings.RollbackSectionName)...)
+}
+
+func (m *DataMigration) updatesApplied(updatedCount int) {
+	m.updates = []mongo.WriteModel{}
+	m.updatedCount += updatedCount
+}
+
+func (m *DataMigration) completeUpdates() bool {
+	capacity := m.settings.capacity
+	if capacity != nil {
+		stats := m.GetStats()
+
+		percent := (float64(stats.Fetched) * float64(100)) / float64(*capacity)
+
+		log.Printf("processed %.0f %% of %d records and applied %d changes", percent, *capacity, stats.Applied)
```

Comment:
Would we want a log message for when `capacity == nil`, too? Just without the capacity and percent?